crates/thermocouple-converter/src/lib.rs

```diff
@@ -1,117 +1,131 @@
 pub const TYPE_K_COEF: [[f64; 10]; 3] = [
-    [
-        0.0,
-        25.173462,
-        -1.1662878,
-        -1.0833638,
-        -0.89773540,
-        -0.37342377,
-        -0.086632643,
-        -0.010450598,
-        -0.00051920577,
-        0.0,
+    [                   // Coefficients for -5.891 <= voltage <= 0.0
+        0.0000000E+00,
+        2.5173462E+01,
+       -1.1662878E+00,
+       -1.0833638E+00,
+       -8.9773540E-01,
+       -3.7342377E-01,
+       -8.6632643E-02,
+       -1.0450598E-02,
+       -5.1920577E-04,
+        0.0000000E+00,
     ],
-    [
-        0.0,
-        25.08355,
-        0.07860106,
-        -0.2503131,
-        0.08315270,
-        -0.01228034,
-        0.0009804036,
-        -0.00004413030,
-        0.000001057734,
-        -0.00000001052755,
+    [                   // Coefficients for 0.0 <= voltage <= 20.644
+        0.000000E+00,
+        2.508355E+01,
+        7.860106E-02,
+       -2.503131E-01,
+        8.315270E-02,
+       -1.228034E-02,
+        9.804036E-04,
+       -4.413030E-05,
+        1.057734E-06,
+       -1.052755E-08,
     ],
-    [
-        -131.8058,
-        48.30222,
-        -1.646031,
-        0.05464731,
-        -0.0009650715,
-        0.000008802193,
-        -0.00000003110810,
-        0.0,
-        0.0,
-        0.0,
+    [                   // Coefficients for 20.644 <= voltage <= 54.886
+       -1.318058E+02,
+        4.830222E+01,
+       -1.646031E+00,
+        5.464731E-02,
+       -9.650715E-04,
+        8.802193E-06,
+       -3.110810E-08,
+        0.000000E+00,
+        0.000000E+00,
+        0.000000E+00,
     ],
 ];
 
-///function for power
-/// in: base: f64, exp: i32
-/// out: result: f64
-fn pow(base: f64, exp: i32) -> f64 {
-    let mut result = 1.0;
-    if exp < 0 {
-        result = 1.0 / pow(base, -exp);
-    } else {
-        for _ in 0..exp {
-            result *= base;
-        }
-    }
-    return result;
+/// Converts a 32-bit ADC reading to a temperature in celsius.
+pub fn adc_to_celsius(adc_reading: i32) -> f64 {
+    voltage_to_celsius(adc_to_voltage(adc_reading))
 }
 
-///function to calculate the conversion between voltage to celsius from the thermocoupler.
-///in: voltage: f64
-///out: celsius: f64
+/// Converts a 32-bit ADC reading to a voltage.
+pub fn adc_to_voltage(adc_reading: i32) -> f64 {
+    const REFERENCE_VOLTAGE: f64 = 5.0;
+    const MAX_ADC_VALUE: f64 = 4_294_967_296.0;
+    const V_SCALE: f64 = REFERENCE_VOLTAGE / MAX_ADC_VALUE; 
+
+    adc_reading as f64 * V_SCALE
+}
+
+/// Converts voltage to celsius for type K thermocouples.
 pub fn voltage_to_celsius(voltage: f64) -> f64 {
-    //define variables
+    return match voltage {
+        -5.891..=0.0 => calc_temp_exponential(voltage, &TYPE_K_COEF[0]),
+        0.0..=20.644 => calc_temp_exponential(voltage, &TYPE_K_COEF[1]),
+        20.644..=54.886 => calc_temp_exponential(voltage, &TYPE_K_COEF[2]),
```

Comment:
I had to use ..= for inclusive range here for the tests that were written to pass. I don't like the double inclusion range (I know that if its = it will go with the first branch matched), but something feels off about both being included. If you think it's fine, this seems to work, so oh well.

Reply:
I see what you mean, should be fine just for the tests! :+1: :test_tube: